### ISSUE TITLE
Soucre code of Ensemble models

### ISSUE THREAD

**Issue by thaondc** [polimi-ispl/icpr2020dfdc#65]
Hi @CrohnEngineer,
I followed your tutorial, I'm training ensemble model but I can not find your code to combine different networks in this repository ( Ex: combine EfficientNet B4 and EfficientNet AutoAtt B4). Can you help me to find them?

**Comment by thaondc** [comment]
Hi @CrohnEngineer,

I have refered the paper, the code and the kaggle link that you sent. And I have some question. Can you help me to anwser?
1/ In the paper, section I, it said that: "Specifically, we consider the strong hardware and time constraints imposed by the DFDC [18]. This means that the proposed solution must be able to analyze **4000 videos** in less than 9 hours using at most a single NVIDIA P100 GPU. Moreover, the trained models must occupy less than 1GB of disk space". I think **4000 videos** mentioned is public test dataset on Kaggle. But the public test dataset on Kaggle has only 400 videos.
I want to ask you which dataset is "**4000 videos**" ?
2/ In the paper,in section IV C, it said that: "for the end-to-end training, we either train for a maximum of 20k iterations, indicating as iteration the processing of a batch of 32 faces (16 real, 16 fake) taken randomly and evenly across all the videos of the train split, or until reaching a plateau on the validation loss. Validation of the model in this context is performed every 500 training iterations, on 6000 samples taken again evenly and randomly across all videos of the validation set. The initial learning rate is reduced of a 0.1 factor if the validation loss does not decrease after 10 validation routines (5000 training iterations), and the training is stopped when we reach a minimum learning rate of \[1 × 10^{-10}\];"
I want to asked you 
3/ In your dicussion on [kaggle](https://www.kaggle.com/c/deepfake-detection-challenge/discussion/145841), you said that: "At inference time, we considered **72 frames** per video and looked at all the faces found by Blazeface, keeping only those with a score above a certain threshold.". But in the paper, you have chooses 32 frames.
 How many frame do you take per video?

**Comment by thaondc** [comment]
Hi @CrohnEngineer,

I have referred to the paper, the code and the Kaggle discussion of your team. And I had some questions. May I answer some questions? 

1/ In the paper, section I, it said that: "Specifically, we consider the strong hardware and time constraints imposed by the DFDC [18]. This means that the proposed solution must be able to analyze 4000 videos in less than 9 hours using at most a single NVIDIA P100 GPU. Moreover, the trained models must occupy less than 1GB of disk space". I think "**4000 videos**" mentioned is public test dataset on Kaggle. But the public test dataset on Kaggle has only 400 videos.
May I ask you which dataset is "**4000 videos**"?
2/ In the paper, in section IV.C, it said that: "for the end-to-end training, we either train for a maximum of 20k iterations, indicating as iteration the processing of a batch of 32 faces (16 real, 16 fake) taken randomly and evenly across all the videos of the train split, or until reaching a plateau on the validation loss. Validation of the model in this context is performed every 500 training iterations, on 6000 samples taken again evenly and randomly across all videos of the validation set. The initial learning rate is reduced of a 0.1 factor if the validation loss does not decrease after 10 validation routines (5000 training iterations), and the training is stopped when we reach a minimum learning rate of 1 × 10^(-10);"
Does it mean you trained them for a routine (not an entire epoch) instead of an epoch? Did you take 6000 samples for each validation routine? May I ask you how many routines and iterations you trained? 
3/ In your discussion on Kaggle, you said that: "At inference time, we considered 72 frames per video and looked at all the faces found by Blazeface, keeping only those with a score above a certain threshold.". But in the paper, you have chosen 32 frames. May I ask you how many frames did you take per video?

Thank you.

**Comment by nicobonne** [comment]
Hi @thaondc, 
1) The 4000 videos dataset was the *private* Kaggle test set, not the public. We didn't have access to this dataset during the competition, I think that now those videos are part of the [full DFDC](https://ai.facebook.com/datasets/dfdc/?ref=hackernoon.com).
2) As stated in the paper, an entire epoch was too long to finish in a reasonable time. We instead trained looking for a validation plateau on a random 6k validation samples, or stopping after 20k training batches.
3) We trained on 32 frames but at inference time the number of frames was choosen to match the time/memory constraints of the competition's platform. We saw that with 72 frames per video we did not saturate the NVIDIA P100 memory and we could process all the 4k videos in less than 9h, so we went for it.

**Comment by thaondc** [comment]
Hi @nicobonne,

Thank you for your answer. But I still have some questions. Can you help me to answer? 

2. "We instead trained looking for a validation plateau on a random 6k validation samples, or stopping after 20k training batches." Does it mean you used 6k random validation samples for a validation routine? Besides, I trained many "routines" but the model is not converging, which mean the learning rate is not less than 10^(-10). Can I ask you how many average routines did you train?

**Comment by CrohnEngineer** [comment]
Hey @thaondc ,

> Does it mean you used 6k random validation samples for a validation routine? 

Yes that's true, with the samples taken from videos of the folders 35 to 40 of the train split.  
The actual number of batches (and therefore iterations) is lower depending on the batch dimension of course.

> Besides, I trained many "routines" but the model is not converging, which means the learning rate is not less than 10^(-10). Can I ask you how many average routines did you train?

Usually the models overfitted long before the 20k batch iterations, around 10k or so they converged.  
A small tip: for "convergence", we mean that the network's validation loss has not changed significantly over the last iterations.
If I were you, I would look manually at the validation loss: there might be small fluctuations that change the value of the min loss reached, but that are not really significant in order of magnitude. 
Those fluctuations might cause the training script not to stop or reduce the learning rate, even though the network has already reached good validation loss values.
Hope this help!
Bests,

Edoardo

**Comment by thaondc** [comment]
Thank you so much for your help, @nicobonne  @CrohnEngineer.
I will try to do it again.

**Comment by thaondc** [comment]
Hi @CrohnEngineer,

I read your source code and try to create another version for myself. But I have some problems, I don't understand these [code lines](https://github.com/polimi-ispl/icpr2020dfdc/blob/master/train_binclass.py#L336), in train_binclass.py, from line 336 to 347, what did they mean? And the function: [tb_attention()](https://github.com/polimi-ispl/icpr2020dfdc/blob/master/train_binclass.py#L371), in train_binclass.py, in line 371 what does it do? Can you help me to answer? 

Thanks.

**Comment by CrohnEngineer** [comment]
Hey @thaondc ,

The function tb_attention() logs the attention masks computed by EfficientNetAutoAtt models on Tensorboard for visualization purposes.  
Lines from 336 to 347 simple compute the attention masks for a couple of frames (one REAL and one FAKE), which are then logged using tb_attention(). In this way you can check how the attention masks vary during the training of the EfficientNetAutoAtt models.  
Nothing more :)  
Glad you're digging the code and making your own version! Good luck with your experiments!
Bests,

Edoardo

**Comment by thaondc** [comment]
Thank you @CrohnEngineer. After I received your response, I have started training my own model. But I had a question, did you use pre-trained weights? For example, did you use EfficientNet's weights trained on ImageNet?

**Comment by CrohnEngineer** [comment]
Yes we did use pre-trained weights on ImageNet!  
We tried also training from scratch the models, but the results were not as good as the ones obtained with models pre-trained on ImageNet.  
This is an interesting topic of discussion as all the top-place solutions relied on pre-trained models, but it is still not clear how pre-training on ImageNet can provide an advantage from a forensic perspective.

**Comment by thaondc** [comment]
Thank you.

**Comment by thaondc** [comment]
Hi @CrohnEngineer,
 I am trying to train a model on the FaceForensics++ dataset. In this dataset, you selected 720 videos for training, 140 for
validation, and 140 for the test from the pool of original sequences. Did you take the videos that are ordered in the data frame or take them randomly?
![image](https://user-images.githubusercontent.com/69307085/158540398-4d38c8f4-c3cc-4f17-a652-bc7f863e69c8.png)

**Comment by CrohnEngineer** [comment]
Hey @thaondc ,

1. We first shuffled randomly the rows of the DataFrame containing the info about the FF++ videos. We then took only the originals, put 720 in the training set, 140 in the validation and 140 in the test set. You can see it in the [splits.py](https://github.com/polimi-ispl/icpr2020dfdc/blob/master/isplutils/split.py) file starting from line 54 to line line 61;
2. For each original video in the various splits, we took a real version from YouTube, and then all the fake versions. In [splits.py](https://github.com/polimi-ispl/icpr2020dfdc/blob/master/isplutils/split.py) these are the lines from 62 to 69.

Hope this helps!
Bests,
